### ISSUE TITLE
Ubuntu Focal Virtual Image stuck at grub stage

### ISSUE THREAD

**Issue by kagehisa** [OSInside/kiwi#2236]
## Problem description
I've build a Image from the virtual profile from the following OSInside [description](https://github.com/OSInside/kiwi-descriptions/tree/master/ubuntu/x86_64/ubuntu-focal). The build process works fine besides the already mentioned proxy issues. 
The Image is created and I convert it to an vhdx image via `qemu-img convert -f raw Ubuntu-20.04_appliance.x86_64-1.16.4.raw -O vhdx ubuntu-test.vhdx ` for use on our Hyper-V plattform.
This works fine, but as soon as I try to start the Image it gets stuck at the grub cmd shell. 

## Expected behaviour
The Image boots completely or at least beyond the grub stage.

## Steps to reproduce the behaviour
- Checkout OSInside KIWI descriptions from Github
- Build the appliance with the profile set to `Virtual`
- Optional: Convert to VHDX
- Start a VM with the generated Image

## OS and Software information

* KIWI version: 9.24.55
* Operating system host version: SLES 15 SP4
* Operating system target version: Ubuntu Focal
* Open Build Service version (N/A if not using OBS): N/A
* Koji version (N/A if not using Koji): N/A

**Comment by schaefi** [comment]
Before we dive deeper can you check if it the .raw disk image works on qemu:

```
qemu-kvm -m 4096 -hda Ubuntu-20.04_appliance.x86_64-1.16.4.raw
```

* If it is also stuck there, can you share the error message you got from grub ?
* Did you modify the image description in some way, or could I reproduce the issue on my side by just rebuilding from the description from the kiwi-descriptions git ?

Thanks

**Comment by kagehisa** [comment]
Hi, thanks for the quick reply. 
The only changes I made are in the `repository` and `users` section. Since OBS links and our company proxy don't mix (Or kiwi ignores the no_proxy for translated OBS links).
The changes are: 
```
    <users>
        <user password="$6$h3rFwlpa/W9gERyl$.BUKxZ2ziL8eJzu1uLasfKlx6AA.G0NDqLEMfzirR4OMrebC3Wd13QdviezgHYmsDKkGf9vcxdIUv03br97wY." home="/root" name="root" groups="root"/>
    </users>
    <repository type="apt-deb" repository_gpgcheck="false" imageinclude="true">
        <source path="http://download.opensuse.org/repositories/Virtualization:/Appliances:/Builder/xUbuntu_20.04"/>
    </repository>
    <repository type="apt-deb" alias="Focal-security" distribution="focal-security" components="main multiverse restricted universe" repository_gpgcheck="false">
        <source path="http://packages.oth-regensburg.de/ubuntu/"/>
    </repository>
    <repository type="apt-deb" alias="Focal-updates" distribution="focal-updates" components="main multiverse restricted universe" repository_gpgcheck="false">
        <source path="http://packages.oth-regensburg.de/ubuntu/"/>
    </repository>
    <repository type="apt-deb" alias="Focal" distribution="focal" components="main multiverse restricted universe" repository_gpgcheck="false">
        <source path="http://packages.oth-regensburg.de/ubuntu/"/>
    </repository>
``` 
The other thing I did differently was to inject an http and https proxy entry to the apt conf that kiwi uses during the prepare phase (See: [issue #2234](https://github.com/OSInside/kiwi/issues/2234#issue-1534478832).



> Before we dive deeper can you check if it the .raw disk image works on qemu:
>
> `qemu-kvm -m 4096 -hda Ubuntu-20.04_appliance.x86_64-1.16.4.raw`

This works just fine. It boots up and I can login.

When I convert the raw image to a vhdx for hyper-v the conversion works fine.
```
qemu-img convert -f raw Ubuntu-20.04_appliance.x86_64-1.16.4.raw -O vhdx ubuntu-hv-test.vhdx
```

But when I try to boot it on Hyper-V I immediately get thrown into the GRUB Shell, no error (may be consumed by HyperV).
![grub_msg](https://user-images.githubusercontent.com/37015428/213166713-6363045c-757e-427e-91a8-73375dda1647.PNG)

Grub seems to see the partitions alright, but from here on I'm lost. 

![grub_ls](https://user-images.githubusercontent.com/37015428/213166887-be0c6219-20d5-4ccc-b4c3-881a5c2eb040.PNG)


I'm thankful for every hint. I used the SLES provided JeOS descriptions for building SLES based JeOS IMages for HyperV and that worked just fine. Am I missing something?

Thanks again for the immediate reply

**Comment by kagehisa** [comment]
Ok, when I specifiy the rootdisl manually via the grub shell I can get it to boot with:
```
grub> set root=(hd0,3)
grub> linux /boot/vmlinuz-x.y.z-generic root=/dev/sda3
grub> initrd /boot/initrd.img-x.y.z-generic
grub> boot
```

But what can I do that grub is able to find the root filesystem on its own? Am I missing something?

**Comment by schaefi** [comment]
Hmm, you get dropped into the grub shell because the earlyboot script builtin to the grub core module was not able to read the grub config file (configfile command in the grub shell).

Question is why. So how does hyperV boot ? BIOS mode or EFI mode ?

You said in qemu it works but I guess you booted in BIOS mode. You can simulate the EFI boot via

```
qemu-kvm -bios /usr/share/qemu/ovmf-x86_64-ms.bin -hda ubuntu-hv-test.vhdx
```

Does that drop you into the same error ?

Also in your image build log file can you check if there is a "grub2-mkimage" or "grub-mkimage" as part of your image build process ?

**Comment by schaefi** [comment]
I tried to reproduce the issue but things works in my qemu tests (no HyperV here). Here is what I did

```
kiwi-ng --profile Virtual system boxbuild --box ubuntu -- --description ~ms/Project/kiwi-descriptions/ubuntu/x86_64/ubuntu-focal/ --target-dir /tmp/myfocal
```

just built it without any modifications. I understand you had to do some because of the repo access. However, maybe you can double check if the same packages were used ?

Next I converted:

```
qemu-img convert -f raw Ubuntu-20.04_appliance.x86_64-1.16.4-0.raw -O vhdx foo.vhdx
```

Next I tested:

```
# EFI
qemu-kvm -bios /usr/share/qemu/ovmf-x86_64-ms.bin -hda foo.vhdx

# BIOS
qemu-kvm  -hda foo.vhdx
```

They all work.

Kinda strange this issue

**Comment by kagehisa** [comment]
> Question is why. So how does hyperV boot ? BIOS mode or EFI mode ?

It boots in EFI mode. I tried switching to a gen1 vm (bios mode), and it works.

> You said in qemu it works but I guess you booted in BIOS mode. You can simulate the EFI boot via

> `qemu-kvm -bios /usr/share/qemu/ovmf-x86_64-ms.bin -hda ubuntu-hv-test.vhdx`
> Does that drop you into the same error ?

Same bahavior for the raw and vhdx converted image as on HyperV. I get thrown into the grub shell.

>Also in your image build log file can you check if there is a "grub2-mkimage" or "grub-mkimage" as part of your image build process ?

To be honest, none of them! 

All I get is:
```
"grub2-install": in paths "/home/manji/KIWI/build/usr/sbin:/home/manji/KIWI/build/usr/bin:/home/manji/KIWI/build/sbin:/home/manji/KIWI/build/bin:/home/manji/KIWI/build/usr/local/bin:/home/manji/KIWI/build/usr/local/sbin" exists: "False" mode match: not checked
``` 
and
``` 
"grub2-mkconfig": in paths "/home/manji/KIWI/build/usr/sbin:/home/manji/KIWI/build/usr/bin:/home/manji/KIWI/build/sbin:/home/manji/KIWI/build/bin:/home/manji/KIWI/build/usr/local/bin:/home/manji/KIWI/build/usr/local/sbin" exists: "False" mode match: not checked
EXEC: [grep -q GRUB_USE_LINUXEFI /home/manji/KIWI/build/usr/sbin/grub-mkconfig]
"grub2-mkconfig": in paths "/home/manji/KIWI/build/usr/sbin:/home/manji/KIWI/build/usr/bin:/home/manji/KIWI/build/sbin:/home/manji/KIWI/build/bin:/home/manji/KIWI/build/usr/local/bin:/home/manji/KIWI/build/usr/local/sbin" exists: "False" mode match: not checked
EXEC: [grep -q GRUB_ENABLE_BLSCFG /home/manji/KIWI/build/usr/sbin/grub-mkconfig]
``` 

I will try the steps in your second post as soon as I get my regular buildhost back and I figured out how to use boxbuild. 
Thank you for taking your time to help me with this. It really is strange...

**Comment by kagehisa** [comment]
Ok this Issue can be closed. Since I work on an SLES 15 SP4 system I thought I could use the OBS KIWI repository for SLES 15 SP3 to install KIWI (see [my other issue](https://github.com/OSInside/kiwi/issues/2240#issue-1553057720)). But it seems between SLES 15 SP3 and SP4 there was a major cleanup regarding python modules and repositories. I was able to install and use KIWI with the 15 SP3 repository but it seems some python modules got renamed or are incompatible with the actually needed ones. 

I removed the kiwi-ng package and the dependencies that where installed with it. Afterwards I installed kiwi-ng via pip and rebuild everything. The raw image and the manually converted vhdx image are able to boot on hyper-v type2 (efi) and qemu-kvm with the above mentioned efi option. 

Thank you so much for helping me with this, even though the fault was my somehow weird build installation. 

But I still wonder why JeOS build worked and only the Ubuntu build was affected by it.
